### DISCUSSION TITLE
谈 Golang 中的字符串和字节数组 · /golang-string

### DISCUSSION THREAD

**Discussion by draveness** [draveness/blog-comments#555] [Announcements]
https://draveness.me/golang-string

**Comment by jobryant** [comment]
@Draveness 你好，想了解下你绘制流程图使用的什么工具，真的非常好看。

**Comment by yizibi** [comment]
> @jobryant
> @Draveness 你好，想了解下你绘制流程图使用的什么工具，真的非常好看。

我用这个在线的，挺好用的，[你试试](http://processon.com)

**Comment by crny** [comment]
UTD-8？不是UTF-8?

**Comment by arkrz** [comment]
您好，文中说：
>标准字符串中不能出现换行符号 \n

但实际上是可以有的，[例如这样](https://play.golang.org/p/WLZkhjPqQ1Z)。
但引用的源码中确实有对 \n 的特殊处理，不知为何……

**Comment by draveness** [comment]
> @ThomasZN
> 您好，文中说：
> >标准字符串中不能出现换行符号 \n
> 
> 但实际上是可以有的，[例如这样](https://play.golang.org/p/WLZkhjPqQ1Z)。
> 但引用的源码中确实有对 \n 的特殊处理，不知为何……

我的意思其实是不能使用如下的方式创建字符串

```go
s := "123
456"
```

**Comment by arkrz** [comment]
> @draveness
> > @ThomasZN
> > 您好，文中说：
> > >标准字符串中不能出现换行符号 \n
> > 
> > 但实际上是可以有的，[例如这样](https://play.golang.org/p/WLZkhjPqQ1Z)。
> > 但引用的源码中确实有对 \n 的特殊处理，不知为何……
> 
> 我的意思其实是不能使用如下的方式创建字符串
> 
> ```go
> s := "123
> 456"
> ```

噢这样，好的:)

**Comment by draveness** [comment]
> @ThomasZN

已经对这里有歧义的段落简单进行了修改 :smile:

**Comment by arkrz** [comment]
还有一个疑问：
```
package main

import (
	"fmt"
)

func main() {
	s1 := "string"
	s2 := s1
	s3 := s1[3:]
	fmt.Println(&s1, &s2, &s3)
}

```
代码中，s1 s2 s3指向的底层字节数组是同一个吧？

**Comment by arkrz** [comment]
> @draveness
> > @ThomasZN
> 
> 这个问题非常好，我之前也没还有想过，编译了一下这段代码获得下面的汇编：
> 
> ```go
> LEAQ	go.string."string"(SB), AX
> PCDATA	$2, $0
> MOVQ	AX, (SP)
> MOVQ	$6, 8(SP)
> CALL	runtime.convTstring(SB)
> PCDATA	$2, $1
> MOVQ	16(SP), AX
> PCDATA	$2, $0
> PCDATA	$0, $1
> MOVQ	AX, ""..autotmp_35+72(SP)
> PCDATA	$2, $2
> LEAQ	go.string."string"(SB), CX
> PCDATA	$2, $0
> MOVQ	CX, (SP)
> MOVQ	$6, 8(SP)
> CALL	runtime.convTstring(SB)
> PCDATA	$2, $1
> MOVQ	16(SP), AX
> PCDATA	$2, $0
> PCDATA	$0, $2
> MOVQ	AX, ""..autotmp_36+64(SP)
> PCDATA	$2, $2
> LEAQ	go.string."string"+3(SB), CX
> PCDATA	$2, $0
> MOVQ	CX, (SP)
> MOVQ	$3, 8(SP)
> CALL	runtime.convTstring(SB)
> PCDATA	$2, $1
> MOVQ	16(SP), AX
> ```
> 
> 我们会发现他们都会调用下面的 convTstring 函数：
> 
> ```go
> func convTstring(val string) (x unsafe.Pointer) {
> 	if val == "" {
> 		x = unsafe.Pointer(&zeroVal[0])
> 	} else {
> 		x = mallocgc(unsafe.Sizeof(val), stringType, true)
> 		*(*string)(x) = val
> 	}
> 	return
> }
> ```
> 
> 所以**三个**赋值表达式都会调用上述方法赋值字符串的内容，所以底层的数组和地址也是不同的。

但是，我刚试了一下，发现底层字符数组是一样的：
```
package main

import (
	"fmt"
	"unsafe"
)

func main() {
	s1 := "string"
	s2 := s1
	s3 := s1[3:]
	fmt.Println(&s1, &s2, &s3)

	printAddr := func(sp *string) {
		unsafePtr := unsafe.Pointer(sp)
		ptr := (*uintptr)(unsafePtr)
		fmt.Println(*ptr, *(*string)(unsafe.Pointer(ptr)))
	}

	printAddr(&s1)
	printAddr(&s2)
	printAddr(&s3)
}
```
您可以试下。还有 不知能否要一下社交平台的联系方式呢~大佬好强的。

**Comment by draveness** [comment]
@ThomasZN 刚到家，我重新编译了这段代码，发现结果完全不同了，非常诡异，可能是我之前姿势的问题，你给我发个邮件吧 i@draveness.me

```go
LEAQ	go.string."string"(SB), AX
MOVQ	AX, "".s1+40(SP) // 初始化 s1
MOVQ	$6, "".s1+48(SP) // 长度 6
PCDATA	$2, $0
PCDATA	$0, $2
MOVQ	AX, "".s2+24(SP) // 发生了复制，指针与 s1 指向同样的数据
MOVQ	$6, "".s2+32(SP) // 长度 6
PCDATA	$2, $1
MOVQ	"".s1+40(SP), AX 
MOVQ	"".s1+48(SP), CX
CMPQ	CX, $3
JCS	192 // panic 如果 3 越界了
PCDATA	$0, $3
ADDQ	$-3, CX 
MOVQ	CX, DX
NEGQ	CX
SARQ	$63, CX
ANDQ	$3, CX // 指针 offset
ADDQ	CX, AX // 修改指针 offset
PCDATA	$2, $0
MOVQ	AX, "".s3+8(SP) // s1 指针 + 3 赋值给 s3
MOVQ	DX, "".s3+16(SP) // 长度 3
```

> 为了避免后面读者误解，我删除了上面的这段代码

**Comment by xgp01** [comment]
请问一下将字符串和byte做类型转换会涉及到底层数组的拷贝吗, 比如这样的代码
```go
s := "hello" // string
b := *(*[]byte)(unsafe.Pointer(&s)) // byte
```

**Comment by draveness** [comment]
> 请问一下将字符串和byte做类型转换会涉及到底层数组的拷贝吗, 比如这样的代码
> 
> ```go
> s := "hello" // string
> b := *(*[]byte)(unsafe.Pointer(&s)) // byte
> ```

你可以试一下，从编译出的汇编代码就能看出到底做没做拷贝

**Comment by xgp01** [comment]
> @draveness
> > 请问一下将字符串和byte做类型转换会涉及到底层数组的拷贝吗, 比如这样的代码
> > 
> > ```go
> > s := "hello" // string
> > b := *(*[]byte)(unsafe.Pointer(&s)) // byte
> > ```
> 
> 你可以试一下，从编译出的汇编代码就能看出到底做没做拷贝

```x86asm
PCDATA	$2, $1
PCDATA	$0, $1
LEAQ	go.string."hello"(SB), AX
PCDATA	$2, $0
MOVQ	AX, "".s+32(SP)
MOVQ	$5, "".s+40(SP)
MOVQ	"".s+48(SP), CX
MOVQ	CX, "".b.cap+24(SP)
```
我试了一下, 看起来应该是没有发生拷贝吧 :)

**Comment by draveness** [comment]
> @XieGP
> > @draveness
> > > 请问一下将字符串和byte做类型转换会涉及到底层数组的拷贝吗, 比如这样的代码
> > > 
> > > ```go
> > > s := "hello" // string
> > > b := *(*[]byte)(unsafe.Pointer(&s)) // byte
> > > ```
> > 
> > 你可以试一下，从编译出的汇编代码就能看出到底做没做拷贝
> 
> ```x86asm
> PCDATA	$2, $1
> PCDATA	$0, $1
> LEAQ	go.string."hello"(SB), AX
> PCDATA	$2, $0
> MOVQ	AX, "".s+32(SP)
> MOVQ	$5, "".s+40(SP)
> MOVQ	"".s+48(SP), CX
> MOVQ	CX, "".b.cap+24(SP)
> ```
> 我试了一下, 看起来应该是没有发生拷贝吧 :)

从生成的汇编语言来看，值拷贝了字符串的地址

**Comment by sadwxqezc** [comment]
可以顺带介绍下string immutable的意义

---

*2020-10-19 UPDATES: 增加了新的章节介绍不可变字符串的意义*

**Comment by draveness** [comment]
> 可以顺带介绍下string immutable的意义

可以的，后面 Review 的时候会加上

**Comment by xiao7737** [comment]
> 而反引号声明的字符串就可以摆脱单行的限制，因为双引号不再负责标记字符串的开始和结束，我们可以在字符串内部直接使用 "，在遇到需要手写 JSON 或者其他复杂数据格式的场景下非常方便

博主这里是写错了吧：“因为双引号不再负责标记字符串的开始和结束”   应该是反引号

**Comment by draveness** [comment]
> > 而反引号声明的字符串就可以摆脱单行的限制，因为双引号不再负责标记字符串的开始和结束，我们可以在字符串内部直接使用 "，在遇到需要手写 JSON 或者其他复杂数据格式的场景下非常方便
> 
> 博主这里是写错了吧：“因为双引号不再负责标记字符串的开始和结束” 应该是反引号

没错啊，这里说的是反引号中的双引号

**Comment by xiao7737** [comment]
嗦嘎


发自我的iPhone

------------------ 原始邮件 ------------------
发件人: Draven <notifications@github.com&gt;
发送时间: 2020年7月16日 11:11
收件人: draveness/blog-comments <blog-comments@noreply.github.com&gt;
抄送: xiaoxc <773708486@qq.com&gt;, Comment <comment@noreply.github.com&gt;
主题: 回复：[draveness/blog-comments] 谈 Golang 中的字符串和字节数组 · /golang-string (#139)





   
而反引号声明的字符串就可以摆脱单行的限制，因为双引号不再负责标记字符串的开始和结束，我们可以在字符串内部直接使用 "，在遇到需要手写 JSON 或者其他复杂数据格式的场景下非常方便
  
博主这里是写错了吧：“因为双引号不再负责标记字符串的开始和结束” 应该是反引号
  
没错啊，这里说的是反引号中的双引号

**Comment by dillanzhou** [comment]
请教一下stringtoslicebyte中buf不为nil的情况是怎么出现的，会在什么语法中用到？

**Comment by draveness** [comment]
> @dillanzhou
> 请教一下stringtoslicebyte中buf不为nil的情况是怎么出现的，会在什么语法中用到？

看一下这个 Commit 你应该就明白了 [cmd/gc: allocate buffers for non-escaping string conversions on stack](https://github.com/golang/go/commit/9568126f350b10163155045445cb149323a2b5c8)

**Comment by 0x1306a94** [comment]
无论是 import 语句中包的路径、结构体中的字段标签还是表达式中的字符串都会使用这个方法将原生字符串中最后的换行符删除并对字符串 Token 进行 Unquote，也就是去掉字符串两遍的引号等无关干扰，还原其本来的面目。

这段中 `也就是去掉字符串两遍的引号等无关干扰` `两遍` 是不是写错了

---

*2021-01-25 UPDATES: 已修复*

**Comment by Mstch** [comment]
> @XieGP
> 请问一下将字符串和byte做类型转换会涉及到底层数组的拷贝吗, 比如这样的代码
> ```go
> s := "hello" // string
> b := *(*[]byte)(unsafe.Pointer(&s)) // byte
> ```

这段代码在访问b.cap的时候应该会fault address

竟然没有fault address,而是得到一个非预期的值

**Comment by zkfopen** [comment]
3.4节 "字符串是它实际上是由字符组成的数组"
是不是可以修改为：
"字符串实际上是由字符组成的数组"

3.4.4节 "这两个情况处理起来都非常简单"
是不是修改为下面更好：
 "这两种情况处理起来都非常简单"

---

*2021-04-01 UPDATES：已修复*

**Comment by sevico** [comment]
if len(s) > 0 && s[0] == '`' {
			s = strings.Replace(s, "\r", "", -1)
		}
这里的意思是只在rawString的情况下才会替换掉所有的\r符号吧，而不是文中说的对所有字符串都替换

**Comment by draveness** [comment]
> @sevico
> if len(s) > 0 && s[0] == '`' {
> 			s = strings.Replace(s, "\r", "", -1)
> 		}
> 这里的意思是只在rawString的情况下才会替换掉所有的\r符号吧，而不是文中说的对所有字符串都替换

你说的是文中那一段有问题？

**Comment by lefeck** [comment]
你好,我想问一下,go语言中struct类型的变量, 在代码编译成汇编语言时，是怎样操作的?

**Comment by lefeck** [comment]
在 string 和 []byte 类型之间反复转换实现修改, 将变量的类型转换成 []byte 后并修改字节数据；如何修改字节数据, 具体举个例子可以吗?

**Comment by wyuhuang** [comment]
> @asveg
> 在 string 和 []byte 类型之间反复转换实现修改, 将变量的类型转换成 []byte 后并修改字节数据；如何修改字节数据, 具体举个例子可以吗?

go中的字节就是uint8, 即修改一个整形数.然后go中的字符串转换成的编码是UTF-8编码的字节数据
str := "aa"
b := []byte(str)
b[0] = 66
str = string(b) // str = "Ba"

**Comment by zyj123** [comment]
cmd/compile/internal/gc.addstr 能帮助我们在编译期间选择合适的函数对字符串进行拼接，该函数会根据带拼接的字符串数量选择不同的逻辑：
发现错别字一枚， 带拼接 -> 待拼接

**Comment by xuguoqiang-ascaler** [comment]
我只想问一下 这个图是怎么画的。。。

**Comment by YCMyTot** [comment]
```
cmd/compile/internal/gc.addstr 能帮助我们在编译期间选择合适的函数对字符串进行拼接，该函数会根据带拼接的字符串数量选择不同的逻辑：
```
这段的 带拼接 -> 待拼接

**Comment by zkfopen** [comment]
已收到您的邮件，谢谢！

**Comment by zhaobuqi-01** [comment]
字符串底层应该是字节切片吧，

**Reply by zhaobuqi-01** [reply]
网上有的说是字节数组有的说是字节切片，但官方博客中说的是string实质上是字节切片

**Comment by zkfopen** [comment]
已收到您的邮件，谢谢！

**Comment by JokerDevops** [comment]
详细分析了字符串与 []byte 类型的关系，我没有明确的看出来字符串和 []byte 之间的关系是什么？只是字符串可以转为 []byte ，然后字符串是一个只读的数组，而[]byte 可是可以读写修改的。最后就是 字符串和 []byte 的内容是相同的，可以通过修改 []byte 的方式进行修改字符串数据，但是修改之后再转为字符串类型不是一样会出现内存copy吗？

**Reply by JokerDevops** [reply]
@draveness

**Comment by zkfopen** [comment]
已收到您的邮件，谢谢！
